File: data_structures/programmers_mock_test.cpp
/*

-- simple compare --

테스트 1 〉	통과 (0.01ms, 4.1MB)
테스트 2 〉	통과 (0.01ms, 4.15MB)
테스트 3 〉	통과 (0.01ms, 4.11MB)
테스트 4 〉	통과 (0.01ms, 4.17MB)
테스트 5 〉	통과 (0.01ms, 4.17MB)
테스트 6 〉	통과 (0.01ms, 4.1MB)
테스트 7 〉	통과 (0.03ms, 4.18MB)
테스트 8 〉	통과 (0.01ms, 4.09MB)
테스트 9 〉	통과 (0.06ms, 4.11MB)
테스트 10 〉	통과 (0.03ms, 4.16MB)
테스트 11 〉	통과 (0.06ms, 4.17MB)
테스트 12 〉	통과 (0.05ms, 4.13MB)
테스트 13 〉	통과 (0.01ms, 3.59MB)
테스트 14 〉	통과 (0.05ms, 4.24MB)
*/

#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> answers) {
    vector<int> answer;
    
    string student1 = "12345";
    string student2 = "21232425";
    string student3 = "3311224455";
    
    int cnt1 = 0, cnt2 = 0, cnt3 = 0;
    
    for (int i = 0; i < answers.size(); i++) {
        if (student1[i%student1.length()] - '0' == answers[i]) cnt1++;
        if (student2[i%student2.length()] - '0' == answers[i]) cnt2++;
        if (student3[i%student3.length()] - '0' == answers[i]) cnt3++;
    }

    if (cnt1 == cnt2 && cnt2 == cnt3) answer = {1,2,3};
    else {
        if (cnt1 >= cnt2) {
            if (cnt1 >= cnt3) answer.push_back(1);
            if (cnt3 >= cnt1) answer.push_back(3);
        }
        if (cnt2 >= cnt1) {
            if (cnt2 >= cnt3) answer.push_back(2);
            if (cnt3 >= cnt2) answer.push_back(3);
        }     
    }
    
    return answer;
}

/*

-- max_element --

테스트 1 〉	통과 (0.01ms, 4.18MB)
테스트 2 〉	통과 (0.01ms, 4.18MB)
테스트 3 〉	통과 (0.01ms, 4.23MB)
테스트 4 〉	통과 (0.01ms, 4.2MB)
테스트 5 〉	통과 (0.01ms, 4.1MB)
테스트 6 〉	통과 (0.01ms, 4.1MB)
테스트 7 〉	통과 (0.06ms, 4.16MB)
테스트 8 〉	통과 (0.02ms, 4.09MB)
테스트 9 〉	통과 (0.09ms, 4.07MB)
테스트 10 〉	통과 (0.05ms, 4.18MB)
테스트 11 〉	통과 (0.10ms, 4.09MB)
테스트 12 〉	통과 (0.09ms, 4.18MB)
테스트 13 〉	통과 (0.01ms, 4.17MB)
테스트 14 〉	통과 (0.04ms, 4.11MB)
*/

#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> solution(vector<int> answers) {
    vector<int> answer;
    vector<int> score(3, 0);
    
    string student1 = "12345";
    string student2 = "21232425";
    string student3 = "3311224455";
    
    for (int i = 0; i < answers.size(); i++) {
        if (student1[i%student1.length()] - '0' == answers[i]) score[0]++;
        if (student2[i%student2.length()] - '0' == answers[i]) score[1]++;
        if (student3[i%student3.length()] - '0' == answers[i]) score[2]++;
    }

    int max_score = *max_element(score.begin(), score.end());
    
    for (int i = 0; i < 3; i++) {
        if (score[i] == max_score) answer.push_back(i+1);
    }
    
    return answer;
}

File: data_structures/programmers_mock_test.java
/*
테스트 1 〉	통과 (0.05ms, 76.3MB)
테스트 2 〉	통과 (0.04ms, 77.4MB)
테스트 3 〉	통과 (0.06ms, 77.7MB)
테스트 4 〉	통과 (0.05ms, 78.8MB)
테스트 5 〉	통과 (0.05ms, 77.3MB)
테스트 6 〉	통과 (0.06ms, 75.4MB)
테스트 7 〉	통과 (0.61ms, 73.5MB)
테스트 8 〉	통과 (0.14ms, 79.5MB)
테스트 9 〉	통과 (0.59ms, 77.8MB)
테스트 10 〉	통과 (0.30ms, 79MB)
테스트 11 〉	통과 (1.00ms, 80.1MB)
테스트 12 〉	통과 (0.90ms, 79.2MB)
테스트 13 〉	통과 (0.10ms, 72.6MB)
테스트 14 〉	통과 (0.59ms, 78.2MB)
*/

import java.util.*;

class Solution {
    public int[] solution(int[] answers) {
        int[] s1 = {1,2,3,4,5};
        int[] s2 = {2,1,2,3,2,4,2,5};
        int[] s3 = {3,3,1,1,2,2,4,4,5,5};
        
        int[] score = new int[3];
        
        for (int i = 0; i < answers.length; i++) {
        	if (s1[i%s1.length] == answers[i]) score[0]++;
        	if (s2[i%s2.length] == answers[i]) score[1]++;
        	if (s3[i%s3.length] == answers[i]) score[2]++;
        }
        
        int maxScore = Math.max(score[0], Math.max(score[1], score[2]));
        
        List<Integer> list = new ArrayList<Integer>();
        for (int i = 0; i < 3; i++) {
        	if (score[i] == maxScore) list.add(i+1);
        }
        
        int[] result = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
        	result[i] = list.get(i);
        }
        
        return result;
    }
}
